Refactor lib to use plonk

CPerezz opened the pull request:
## Adds
- travis basic configfile
- Optimization on the Sponge hash fn to skip useless addition if the width of the messages is the appropiate one to do it. (WIDTH < HADES_WIDTH -1).
- CHANGELOG & CONTRIBUTING files.
## Changes
- Sponge hash refactor with gadget and scalar versions.

## Removes
- Old tests which no longer apply.

Closes #5

I bumped to 0.2, so if this gets merged I'll tag it but probably not release it since it's not finnished. @ZER0 let me know WDYT.

CPerezz committed:
Remove legacy tests

CPerezz committed:
Add basic travis configfile

CPerezz committed:
Update dependencies

We no longer rely on Buletproofs nor Curve25519.
Instead we use dusk-bls12_381 and dusk-plonk as
it's replacements.

CPerezz committed:
Refactor pad to work with variables

Now the function requires the user to provide the padder value
and the eom value which should be the null and identity values
in respect to their types.

- Refactored pad function to be generic but with this different logic.
- Refactored the tests to work with variables and this new logic.

CPerezz committed:
Modify lib structure

CPerezz committed:
Refactor sponge_hash function for Poseidon w/plonk

- Implemented sponge_hash_gadget to work with plonk and
perform the sponge hash technique withing a circuit.
- Added an optimization which depending on the width of the
padded words, allows skipping the null additions.
- Added tests for various width inputs covering all of the
possible cases.

Partially-Closes #5 partially since we need to implement  the merkle
logic.

CPerezz committed:
Add changelog & contributing files

ZER0 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,69 @@
+# Contributing to plonk
```
I guess it should be "Poseidon"

ZER0 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,69 @@
+
+- Keep your pull requests as `Draft` until they are ready to be reviewed.
+
+- Be descriptive in the titles of your PRs and also provide good descriptions about
```
I would also add somewhere a reference to [How to Write a Git Commit Message](https://chris.beams.io/posts/git-commit/)

ZER0 commented on Cargo.toml:
```diff
@@ -1,21 +1,17 @@
-version = "0.1.0"
+version = "0.2.0"
 authors = [
   "zer0 <matteo@dusk.network>"
```
Add yourself to the list of authors, as well as Victor.

CPerezz replied:
Shouldn't we wait for #11 to land to then make the bump?

ZER0 commented on Cargo.toml:
```diff
@@ -1,21 +1,17 @@
-git = "https://github.com/dusk-network/bulletproofs"
-branch = "develop"
-features = ["yoloproofs"]
+hades252 = { git = "https://github.com/dusk-network/hades252", branch = "master" }
```
Which version of Hades are we targeting?

CPerezz replied:
I'm picking "master" since there's no tag or version for the newes't implementation which is using PLONK as constraint system & proof system.

I think that you were going to add the tag once the history was re-written. But I checked and it hasn't been created.

ZER0 commented on src/sponge/pad.rs:
```diff
@@ -1,16 +1,17 @@
+//!
 
-pub(crate) fn pad<T>(messages: &[T], width: usize) -> Vec<T>
+pub(crate) fn pad<T>(messages: &[T], width: usize, pad_value: T, eom: T) -> Vec<T>
```
for consistency maybe we could call it `eom_value` at this point?

ZER0 commented on src/sponge/pad.rs:
```diff
@@ -1,16 +1,17 @@
-use curve25519_dalek::scalar::Scalar;
+//! Padding support for sponge hash
+//!
 
```
Maybe we could be more descriptive about what it does, at least how the `pad_value` and `eom` are used.

CPerezz replied:
Not targetting to have a good documentation atm.

I just put this so I don't get errors due to the flag `#[deny(missing_docs)]`.
I was leaving docs for the end once #11 gets merged.

ZER0 commented on src/sponge/pad.rs:
```diff
@@ -40,10 +42,13 @@ mod tests {
 
-        assert_eq!(&pad(&[two], 3), &[zero, two, one]);
-        assert_eq!(&pad(&[two, three], 3), &[zero, two, three, zero, one, zero]);
+        assert_eq!(&pad(&[two], 3, zero, one), &[zero, two, one]);
```
I would also rename in the tests `zero` and `one` here as `pad_value` and `eom_value` when are used for such purpose, it would be clearer.

ZER0 commented on src/sponge/pad.rs:
```diff
@@ -56,14 +61,35 @@ mod tests {
+        );
+    }
+
+    /* No longer applies.
```
Shall we remove it, then?

CPerezz replied:
Yepp

ZER0 commented on src/sponge/sponge.rs:
```diff
@@ -0,0 +1,203 @@
+//! The `pad` module implements the Sponge's padding algorithm
```
It's a bit unclear to me this comment.

CPerezz replied:
Not targetting to have a good documentation atm.

I just put this so I don't get errors due to the flag #[deny(missing_docs)].
I was leaving docs for the end once #11 gets merged.

ZER0 commented on src/sponge/sponge.rs:
```diff
@@ -0,0 +1,203 @@
+pub fn sponge_hash(messages: &[Scalar]) -> Scalar {
+    let mut strategy = ScalarStrategy::new();
+
+    let mut words = pad(messages, WIDTH, Scalar::zero(), Scalar::one());
```
For readability, I would do something like:
```rust
// Filling the vector returned by `pad` with zero
let pad_value = Scalar::zero();
// one will identify the end of messages
let eom_value = Scalar::one();

let mut words = pad(messages, WIDTH, pad_value, eom_value);
```

Or something like that.

CPerezz replied:
Yep, agree!

ZER0 commented on src/sponge/sponge.rs:
```diff
@@ -0,0 +1,203 @@
+/// The `hash` function takes an arbitrary number of plonk `Variable`s and returns the
+/// hash, using the `Hades` GadgetStragegy
+pub fn sponge_hash_gadget(composer: &mut StandardComposer, messages: &[Variable]) -> Variable {
+    let padder = composer.zero_var;
```
Ah ah ah, I like the fact that you ALREADY did what I suggested in the review for the gadget part. :) So, yeah, be consistent, you can add `padder` and `eom` also for the scalar's one.

CPerezz replied:
Done avobe!

ZER0 commented on src/sponge/sponge.rs:
```diff
@@ -0,0 +1,203 @@
+                    Scalar::zero(),
+                )
+            });
+            let mut strategy = GadgetStrategy::new(composer);
```
At this point you know better than me: since you always pass the same `composer`, do we need to create a new strategy all the time as well? Shouldn't we create it just once after the composer?

CPerezz replied:
Since it is mutably borrowed, I cannot create the strategy at the beggining of the function since then the `&mut Composer` would be already borrowed by the strategy and I will not be able to call it's functions or attributes.

ZER0 commented on src/sponge/sponge.rs:
```diff
@@ -0,0 +1,203 @@
+    fn poseidon_sponge_params(width: usize) -> (Vec<Scalar>, Scalar) {
+        let mut input = vec![Scalar::zero(); width];
+        input.iter_mut().for_each(
+            |s| *s = Scalar::one(), /*Scalar::random(&mut rand::thread_rng())*/
```
What's the comment line stands for

CPerezz replied:
Some testing debuggin I was doing. Mb

ZER0 reviewed (changes requested):
LGTM, I would just change few nits and I've a couple of questions, but it's good overall

CPerezz committed:
Add missing authors to Cargo.toml

CPerezz committed:
Add commiting guide to CONTRIBUTING.md

CPerezz committed:
Rename testing padding values appropiately

As @zer0 suggested, it is more readable to generate
One and Zero as pad_value and eom_value.

CPerezz committed:
Refactor sponge hash test values & padder names

- It is easier now to understand which value is
used as padder and which is used as eom since we've
renamed them and added some comments.

- Refactored the value initializer for sponge hash
tests to provide random Scalars instead of fixed
values.

ZER0 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,69 @@
+
+- Be descriptive in the titles of your PRs and also provide good descriptions about
+which things are being added, changed, fixed or removed also addressing the corresponding
+issue that it is addressing to. See https://chris.beams.io/posts/git-commit/ for more info.
```
For consistency, I would actually wrote:
`See [How to Write a Git Commit Message]( https://chris.beams.io/posts/git-commit/)`

ZER0 commented on src/sponge/pad.rs:
```diff
@@ -30,73 +31,73 @@ where
     fn test_scalar_padding_width_4() {
-        let zero = Scalar::zero();
-        let one = Scalar::one();
+        let pad_value = Scalar::zero();
```
Since you used `padder` and `eom` as identifiers later on, I think we should be consistent and use them in the tests as well. Or viceversa. Whatever you like more.

ZER0 commented on Cargo.toml:
```diff
@@ -1,21 +1,17 @@
-git = "https://github.com/dusk-network/bulletproofs"
-branch = "develop"
-features = ["yoloproofs"]
+hades252 = { git = "https://github.com/dusk-network/hades252", branch = "master" }
```
As discussed, point to `0.5.0` already. Even if it fails the build on travis it will pass once we add the tag.

ZER0 reviewed (commented):
LGTM, with the few nits addressed.

CPerezz committed:
Consistent padder and eom var names on tests

CPerezz committed:
More explicit ref to "How to Commit page"

CPerezz committed:
Use Hades252 v0.5.0 instead of "master"

Following the path that we're taking to have everything
with correct versioning, we need to migrate from "master"
to the correct version of the git repo of Hades252.
